components/raftstore/src/store/fsm/apply.rs

```diff
@@ -188,7 +216,8 @@ pub enum ExecResult {
     ComputeHash {
         region: Region,
         index: u64,
-        snap: RocksSnapshot,
+        snap: E::Snapshot,
+        _phantom: PhantomData<E>,
```

Comment:
Could we use  `impl Trait` to avoid too many type parameters? Because this result will be send to another thread and it is not need by user request, impl or dyn is enough.

Reply:
It's not possible to use `impl Trait` here because `impl Trait` is not allowed in structures, and its not possible to use `dyn Trait` because `Snapshot` isn't object safe.

`Snapshot` is pretty ugly right now because it carries an engine type parameter. This is at least one reason it isn't object safe.

That type parameter is only needed though because `Snapshot` contains a `get_db` method. This seems pretty extravagant. It might be a simplification to remove that method and pass around `KvEngine` instances where its used, though depending on the use it might not be a simplification.

After some experimenting though, even if `Snapshot` didn't have type parameters, it still has three associated types that need to be concretely specified in order to be turned into an object.

In this particular case, the Snapshot _appears_ to be used only for computing a "region hash" in consistency_check.rs. It might be possible to create an object_safe `SnapshotHasher` trait that sinks all the hashing logic into the engine implementation. That might allow a number of type parameters around this subject to disappear.

That's a pretty big task though. Can I try it in a follow up?

Reply:
Yes. I have pull one PR that remove `get_db` method from `engine_traits::Snapshot`.

Reply:
I'll rework this patch a bit once yours lands. It will at least simplify a few types by not using `E::Snapshot` so often.